lib/xnvme_be_vfio_mem.c

```diff
@@ -84,6 +84,53 @@ xnvme_be_vfio_buf_vtophys(const struct xnvme_dev *XNVME_UNUSED(dev), void *XNVME
 	errno = ENOSYS;
 	return -ENOSYS;
 }
+
+void *
+xnvme_be_vfio_buf_map(const struct xnvme_dev *dev, size_t nbytes, uint64_t *XNVME_UNUSED(phys), void *vaddr)
+{
+	struct xnvme_be_vfio_state *state = (void *)dev->be.state;
+	struct vfio_state *vfio = &state->ctrl->pci.vfio;
+  struct vfio_iommu_mapping *mapping;
+	
+	XNVME_DEBUG("xnvme_be_vfio_buf_map(%p, %p)", dev, vaddr);
+
+	mapping = vfio_iommu_find_mapping(&vfio->iommu, vaddr);
+	if (mapping) {
+		XNVME_DEBUG("Already mapped for %p\n", vaddr);
+		return vaddr;
+	}
+
+	nbytes = ALIGN_UP(nbytes, __VFN_PAGESIZE);
+
+	XNVME_DEBUG("vfio_map_vaddr(%p, %p, %ld, NULL)", vfio, vaddr, nbytes);
+	if (vfio_map_vaddr(vfio, vaddr, nbytes, NULL)) {
+		XNVME_DEBUG("FAILED: vfio_map_vaddr(): %s\n", strerror(errno));
+		return NULL;
+	}
+
+	return vaddr;
+}
+
+void
+xnvme_be_vfio_buf_unmap(const struct xnvme_dev *dev, void *buf)
+{
+	struct xnvme_be_vfio_state *state = (void *)dev->be.state;
+	struct vfio_state *vfio = &state->ctrl->pci.vfio;
+  struct vfio_iommu_mapping *mapping;
+	
+	XNVME_DEBUG("xnvme_be_vfio_buf_unmap(%p, %p)", dev, buf);
+	
+  mapping = vfio_iommu_find_mapping(&vfio->iommu, buf);
+	if (!mapping) {
+			XNVME_DEBUG("no mapping for %p\n", buf);
+		return;
+	}
+
+	if (vfio_unmap_vaddr(vfio, buf)) {
```

Comment:
In libvfn v2.0.0 this function will take an additional `size_t *len` parameter. If not `NULL` it will be set to the length of the mapping being unmapped. This helps the application deallocate the memory if it did so with, say, `mmap()`.